Removed the Material Dialog Box and Added Alert Box from DeckPickerAnalyticsOption

neeldoshii opened the pull request:
## Fixes
* Related #13315  

## Approach
Replaces usages of Deprecated MaterialDialog and adds the Native AlertDialog For Deck Picker Analytics.

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [X] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [X] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

lukstbit commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DeckPickerAnalyticsOptInDialog.kt:
```diff
@@ -17,28 +17,32 @@
+    override fun onCreateDialog(savedInstanceState: Bundle?): AlertDialog {
         super.onCreateDialog(savedInstanceState)
-        return MaterialDialog(requireActivity()).show {
+        return AlertDialog.Builder(requireActivity()).apply {
```
We have an extension method _show {}_ which can be used instead of the _apply {}.show()_ combination.

BrayanDSO replied:
Shouldn't this use `create {}` instead?

lukstbit replied:
We don't have an extension _create {}_ that can be used, but we should. There's one extension _create {}_ but that executes the lambda against the created dialog and not the builder.

Also _show {}_ works as well.

david-allison replied:
Agreed. `.create {` should accept a `Builder`

neeldoshii replied:
> `.show{}` calls the `.show()` routine, which depends on a window.decorView and exhibits the dialog on it. Since the method goal is to deliver a `Dialog` object, I believe that `.create` is more appropriate

@BrayanDSO  I have changed to `.create` in my commit can you review it?

david-allison replied:
This should be 

```diff
- .apply {
+ .create {
```

lukstbit commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DeckPickerAnalyticsOptInDialog.kt:
```diff
@@ -17,28 +17,32 @@
             message(R.string.analytics_summ)
-            checkBoxPrompt(R.string.analytics_title, isCheckedDefault = true, onToggle = null)
+            checkBoxPrompt(R.string.analytics_title, isCheckedDefault = true) { checked ->
+                UsageAnalytics.isEnabled = checked
```
This will not work as expected if the user doesn't click on this checkbox before hitting continue, the _UsageAnalytics.isEnabled_ property must be set in the dialog's positive button.

neeldoshii replied:
If the user doesn't clicks on the checkbox then the analytics boolean would be false.

david-allison replied:
This still seems incorrect

neeldoshii replied:
> This still seems incorrect

Thank you David & luskbit for spotting it, there were 2 cases which were messing in my earlier code
1. If the user canceled it (dismissed it) it would even turn on its analytics.
2. The first default checkbox was not making the checkbox bool triggered in `UsageAnalytic`.

I have fixed it in my current commit. 
Thank you

lukstbit reviewed (changes requested):
As always with this type of changes, please provide some before and after pictures with the changed dialog.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DeckPickerAnalyticsOptInDialog.kt:
```diff
@@ -17,28 +17,32 @@
+    override fun onCreateDialog(savedInstanceState: Bundle?): AlertDialog {
         super.onCreateDialog(savedInstanceState)
-        return MaterialDialog(requireActivity()).show {
+        return AlertDialog.Builder(requireActivity()).apply {
```
`show {` should work here

neeldoshii commented:
> As always with this type of changes, please provide some before and after pictures with the changed dialog.

Sorry @lukstbit, I intended this PR as a draft PR as this was in progress and yet to be tested. I am extremely sorry for making this PR ready to review mistakenly and making maintainer code review messed up. 

I am still not sure when does this trigger so I have created a button and triggered from there for testing purpose. I tried changing the build version of the app but still it doesn't triggered up. Since the change is small I did it force triggered using button in local testing.


| Material Dialog | Alert Dialog |
|-----------------|--------------|
| ![](https://github.com/ankidroid/Anki-Android/assets/60827173/3e1a8b60-c808-4d4e-be7b-e5a508ca0984) | ![](https://github.com/ankidroid/Anki-Android/assets/60827173/310fe6be-e5a6-48c3-a2f9-f0117948ab2b) |

### Working Video

**Material Dialog**

https://github.com/ankidroid/Anki-Android/assets/60827173/4f69a7ac-86fc-4c56-82bc-d12e7fe1c315

**Alert Dialog**

https://github.com/ankidroid/Anki-Android/assets/60827173/589593b4-a594-47e4-a48c-efdf086555d8

neeldoshii commented:
> We have an extension method show {} which can be used instead of the apply {}.show() combination.

Should I change it to `show` from `apply` to maintain consistency?

lukstbit commented:
@neeldoshii We added some extension methods that should be used in this migration here. 
First is the _create{}_ extension method that you can find in AlertDialogFacade.kt. so you would write:

```
AlertDialog.Builder.create {
   // the dialog setup code
}
```

We also added a way to get the checkbox prompt from the dialog, _getCheckBoxPrompt()_ that you can use to setup the analytics in the positive button listener without the need of extra variables:

`(it as AlertDialog).getCheckBoxPrompt().isChecked`

neeldoshii commented:
> @neeldoshii We added some extension methods that should be used here. First is the _create{}_ extension method that you can find in AlertDialogFacade.kt. so you would write:
> 
> ```
> AlertDialog.Builder.create {
>    // the dialog setup code
> }
> ```
> 
> We also added a way to get the checkbox prompt from the dialog, _getCheckBoxPrompt()_ that you can use to setup the analytics in the positive button listener without the need of extra variables:
> 
> `(it as AlertDialog).getCheckBoxPrompt().isChecked`

Updated the changes @lukstbit

lukstbit reviewed (approved):
Much better, thanks!

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DeckPickerAnalyticsOptInDialog.kt:
```diff
@@ -17,23 +17,22 @@
+    override fun onCreateDialog(savedInstanceState: Bundle?): AlertDialog {
         super.onCreateDialog(savedInstanceState)
-        return MaterialDialog(requireActivity()).show {
+        return AlertDialog.Builder(requireActivity()).create {
```
This change to `create` is correct, but shouldn't be included in a refactor

The aim of the refactor is to make the change with as little risk as possible

If this were buggy, we'd need to revert the whole refactor rather than the functional change

neeldoshii replied:
> This change to create is correct, but shouldn't be included in a refactor

I am not sure I understood your point here @david-allison .

david-allison replied:
Don't include functional changes in a refactor
